discovery/discovery.go

```diff
@@ -127,51 +127,48 @@ func (o *orchestratorPool) GetOrchestrators(ctx context.Context, numOrchestrator
 	}
 
 	var infos []*net.OrchestratorInfo
-	var suspendedInfos *suspensionQueue
-	var timedOut bool
-	var nbResp int
+	suspendedInfos := newSuspensionQueue()
+	timedOut := false
+	nbResp := 0
+	infoCh := make(chan *net.OrchestratorInfo, numAvailableOrchs)
+	errCh := make(chan error, numAvailableOrchs)
 
-	// try to discover orchestrators until at least 1 is found (with the exponential backoff timout)
-	timeout := getOrchestratorsCutoffTimeout
-	for {
-		ctx, cancel := context.WithTimeout(clog.Clone(context.Background(), ctx), timeout)
-		infoCh := make(chan *net.OrchestratorInfo, numAvailableOrchs)
-		errCh := make(chan error, numAvailableOrchs)
-
-		for _, uri := range uris {
-			go getOrchInfo(ctx, uri, infoCh, errCh)
-		}
+	ctx, cancel := context.WithTimeout(clog.Clone(context.Background(), ctx), maxGetOrchestratorCutoffTimeout)
+	for _, uri := range uris {
+		go getOrchInfo(ctx, uri, infoCh, errCh)
+	}
 
-		timedOut = false
-		suspendedInfos = newSuspensionQueue()
-		nbResp = 0
-		for i := 0; i < numAvailableOrchs && len(infos) < numOrchestrators && !timedOut; i++ {
-			select {
-			case info := <-infoCh:
-				if penalty := suspender.Suspended(info.Transcoder); penalty == 0 {
-					infos = append(infos, info)
-				} else {
-					heap.Push(suspendedInfos, &suspension{info, penalty})
-				}
-				nbResp++
-			case <-errCh:
-				nbResp++
-			case <-ctx.Done():
+	// try to wait for orchestrators until at least 1 is found (with the exponential backoff timout)
+	timeout := getOrchestratorsCutoffTimeout
+	timer := time.NewTimer(timeout)
+
+	for nbResp < numAvailableOrchs && len(infos) < numOrchestrators && !timedOut {
+		select {
+		case info := <-infoCh:
+			if penalty := suspender.Suspended(info.Transcoder); penalty == 0 {
+				infos = append(infos, info)
+			} else {
+				heap.Push(suspendedInfos, &suspension{info, penalty})
+			}
+			nbResp++
+		case <-errCh:
+			nbResp++
+		case <-timer.C:
+			if len(infos) > 0 {
 				timedOut = true
 			}
-		}
-		cancel()
-
-		if len(infos) > 0 {
-			break
-		}
 
-		timeout *= 2
-		if timeout >= maxGetOrchestratorCutoffTimeout {
-			break
+			timer.Reset(timeout)
```

Comment:
Should we move L162-166 first? Otherwise, I think the timer will be set to `getOrchestratorsCutoffTimeout` twice - once when it is initialized on L143 and then again the first time this case is triggered.